compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -40,24 +41,28 @@ impl<'tcx> Visitor<'tcx> for DeduceReadOnly {
         // We're only interested in arguments.
         let Some(param_index) = self.as_param(place.local) else { return };
 
-        let mark_as_mutable = match context {
+        match context {
             // Not mutating, so it's fine.
-            PlaceContext::NonUse(..) => false,
+            PlaceContext::NonUse(..) => {}
             // Dereference is not a mutation.
-            _ if place.is_indirect_first_projection() => false,
+            _ if place.is_indirect_first_projection() => {}
+            // This is a `Drop`. It could disappear at monomorphization, so mark it specially.
+            PlaceContext::MutatingUse(MutatingUseContext::Drop) => {
+                self.read_only[param_index] |= DeducedReadOnlyParam::IF_NO_DROP;
+            }
```

Comment:
It is possible that a type doesn't need to be dropped while its projection does need to be dropped.